File: .tmp_py4013/batch_08_dump/2906_construct_product_matrix.js
// LeetCode 2906 - Construct Product Matrix
// https://leetcode.com/problems/construct-product-matrix/

/**
 * @param {number[][]} grid
 * @return {number[][]}
 */
var constructProductMatrix = function(grid) {
    const mod = 12345;
    const m = grid.length, n = grid[0].length;
    const ans = Array.from({ length: m }, () => Array(n));
    let pref = 1;
    for (let i = 0; i < m; i++)
        for (let j = 0; j < n; j++) {
            ans[i][j] = pref;
            pref = (pref * (grid[i][j] % mod)) % mod;
        }
    let suf = 1;
    for (let i = m - 1; i >= 0; i--)
        for (let j = n - 1; j >= 0; j--) {
            ans[i][j] = (ans[i][j] * suf) % mod;
            suf = (suf * (grid[i][j] % mod)) % mod;
        }
    return ans;
};

File: .tmp_rs4013/rust_check_12/2906_construct_product_matrix.rs
struct Solution;
// LeetCode 2906 - Construct Product Matrix
// https://leetcode.com/problems/construct-product-matrix/

impl Solution {
    pub fn construct_product_matrix(grid: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        const MOD: i64 = 12345;
        let m = grid.len();
        let n = grid[0].len();
        let mut ans = vec![vec![0i32; n]; m];
        let mut pref = 1i64;
        for i in 0..m {
            for j in 0..n {
                ans[i][j] = pref as i32;
                pref = pref * (grid[i][j] as i64 % MOD) % MOD;
            }
        }
        let mut suf = 1i64;
        for i in (0..m).rev() {
            for j in (0..n).rev() {
                ans[i][j] = (ans[i][j] as i64 * suf % MOD) as i32;
                suf = suf * (grid[i][j] as i64 % MOD) % MOD;
            }
        }
        ans
    }
}

fn main() {}
